Reworked skinning implementation

mcdenis opened the pull request:
Reworked the skinning implementation : 

* Subclassed most controls and added a few properties to them so that we can fully customize their appearance in a more standard way (e.g. we can just use the DisabledForeColor property to change the text color when the control is disabled; No more hacky custom drawing needed.) In practice, this also make it possible for controls to better retain their native appearance when using the default skin.
* Skinning of ToolStrips is accomplished using a custom ToolStripRenderer that allows any of its color to be changed while also supporting the system theme, if desired.
* Added several properties to the skin files (i.e. BaseSkin and DarkSkin) and separated properties by control type. This could allows the creation of richer looking skin in the future.
* Removed all custom drawing code from SkinnableForms since each control now handles its own drawing. Instead, the form implements a FormSkin property and values associated with the skin are simply applied to the corresponding control properties.

I have also reworked a bit the layout of some forms and user controls to better leverage Winform's auto layout features. This way, the layout is automatically adjusted to the content which magically fixes most localization and DPI issues.

There is still some work to do in both areas, but I want to see if the changes I made so far are accepted before going any further 😉

Here is a before/after comparison of the main form with both skins:

![dark skin comparison](https://cloud.githubusercontent.com/assets/25092585/24176491/a74d42dc-0e72-11e7-9342-8c60e83a70de.png)
![default skin comparison](https://cloud.githubusercontent.com/assets/25092585/24176492/a7526b4a-0e72-11e7-9cf1-6156c68c6d3f.png)

Jonno12345 commented:
Hey - thanks for all of your hard work on this, looks far better than the quick hacky implementation I put in place! I'll review the code ASAP, but assuming nothing is broken I can't thing of any good reason not to implement, it looks much better like this.

Thanks again!

mcdenis commented:
Thanks you.

I should have also mentioned in the pull request description that I have temporarily "disabled" skinning for
FrmCustomShortcutManagerNew (the form descends from Form instead of SkinnableForm), because I have not yet created a skinnable TabControl. Everything else should work just like before though. I expect to have some free time early next week, so hopefully I will be able to complete the missing control then.

Jonno12345 commented:
Committed a change to the apply skin logic - effectively moved the logic into the controls themselves (utilising an ISkinnableControl interface). 

Let me know your thoughts on this (I did a couple of little cleanups too just for my own readability to fit with the rest of the application).

Otherwise, it's a definite improvement over what I had shoe-horned in, thanks for your hard work!

mcdenis commented:
Thanks for the follow up! Your changes definitely make things cleaner.

Concerning the skinnable TabControl, I finished it yesterday, and it is now ready to be added to the app. I am pretty new to Github, so I am not sure how I should proceed now. Should I create a new branch in my fork based on pr/71, commit the changes and submit a new pull request in pr/71?

Jonno12345 commented:
Just sent you a PR with my amends. Merge those along with your TabControl changes to your 'develop' branch, that should bring this PR up to speed :)

mcdenis commented:
Thanks a lot for help. I have merged your changes and added the skinnable TabControl. 👍

A few notes:

* Since SkinnableTabControl implements ISKinnableControl while also being a container, I have moved the ApplyControlSkin method from SkinnableForm in a separate static class named ContainerUtils so that  SkinnableTabControl can use it to skin its child controls. It would be certainly possible to skin all the child controls in SkinnableForm, but I think it makes more sense if skinnable containers take care themselves of their child controls.
* FrmCustomShortcutManagerNew has minor layout issues with the dark skin, because the latter use a font with different metrics. I will be away for the next few days, but I would be glad to fix these issues next week.

Jonno12345 commented:
Cool, thanks for your help :)

I have one issue I will look into before I commit the PR, resizing the main form is quite jerky which I imagine is due to constant painting. Not sure if it's avoidable (and it's pretty minor) but just to try and understand whats causing it.

mcdenis commented:
I did some test and the jerkiness that you are referring to seems to occur because the columns of the listbox are removed and recreated every time the form is resized.

I made some change in the *mainformjerkiness* branch of my fork that should fix this issue. Could you let me know what you think?

I have also made changes to the math used to calculate the width of columns so that it considers the width of the control's client area, instead of using its full width and guessing the size of the border and the scrollbar.

mcdenis commented:
Added a few commits that polish the layout of FrmCustomShortcutManagerNew (+ minor other changes). It should no longer have the issues with the dark skin that I mentioned previously.

Jonno12345 commented:
Wonderful. Thanks for all your help, much easier on the eyes and a lot easier to maintain :)

mcdenis commented:
Thank you, your application is so fun to work on! I will probably open a new pull request soon with other refinements, including a SkinnableTrackBar 😎

![image](https://cloud.githubusercontent.com/assets/25092585/25028606/adef6fe8-2083-11e7-921f-34bf551e3832.png)
